cmd/server/pactasrv/analysis.go

```diff
@@ -0,0 +1,343 @@
+package pactasrv
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/RMI/pacta/cmd/server/pactasrv/conv"
+	"github.com/RMI/pacta/db"
+	"github.com/RMI/pacta/oapierr"
+	api "github.com/RMI/pacta/openapi/pacta"
+	"github.com/RMI/pacta/pacta"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// (GET /analyses)
+func (s *Server) ListAnalyses(ctx context.Context, request api.ListAnalysesRequestObject) (api.ListAnalysesResponseObject, error) {
+	ownerID, err := s.getUserOwnerID(ctx)
+	if err != nil {
+		return nil, err
+	}
+	as, err := s.DB.AnalysesByOwner(s.DB.NoTxn(ctx), ownerID)
+	if err != nil {
+		return nil, oapierr.Internal("failed to query analyses", zap.Error(err))
+	}
+	items, err := dereference(conv.AnalysesToOAPI(as))
+	if err != nil {
+		return nil, err
+	}
+	return api.ListAnalyses200JSONResponse{Items: items}, nil
+}
+
+// Deletes an analysis (and its artifacts) by ID
+// (DELETE /analysis/{id})
+func (s *Server) DeleteAnalysis(ctx context.Context, request api.DeleteAnalysisRequestObject) (api.DeleteAnalysisResponseObject, error) {
+	id := pacta.AnalysisID(request.Id)
+	_, err := s.checkAnalysisAuthorization(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	blobURIs, err := s.DB.DeleteAnalysis(s.DB.NoTxn(ctx), id)
+	if err != nil {
+		return nil, oapierr.Internal("failed to delete analysis", zap.Error(err))
+	}
+	if err := s.deleteBlobs(ctx, blobURIs...); err != nil {
+		return nil, err
+	}
+	return api.DeleteAnalysis204Response{}, nil
+}
+
+// Returns an analysis by ID
+// (GET /analysis/{id})
+func (s *Server) FindAnalysisById(ctx context.Context, request api.FindAnalysisByIdRequestObject) (api.FindAnalysisByIdResponseObject, error) {
+	a, err := s.checkAnalysisAuthorization(ctx, pacta.AnalysisID(request.Id))
+	if err != nil {
+		return nil, err
+	}
+	if err := s.populateArtifactsInAnalyses(ctx, a); err != nil {
+		return nil, err
+	}
+	if err := s.populateBlobsInAnalysisArtifacts(ctx, a.Artifacts...); err != nil {
+		return nil, err
+	}
+	converted, err := conv.AnalysisToOAPI(a)
+	if err != nil {
+		return nil, err
+	}
+	return api.FindAnalysisById200JSONResponse(*converted), nil
+}
+
+// Updates writable analysis properties
+// (PATCH /analysis/{id})
+func (s *Server) UpdateAnalysis(ctx context.Context, request api.UpdateAnalysisRequestObject) (api.UpdateAnalysisResponseObject, error) {
+	id := pacta.AnalysisID(request.Id)
+	_, err := s.checkAnalysisAuthorization(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	mutations := []db.UpdateAnalysisFn{}
+	if request.Body.Name != nil {
+		mutations = append(mutations, db.SetAnalysisName(*request.Body.Name))
+	}
+	if request.Body.Description != nil {
+		mutations = append(mutations, db.SetAnalysisDescription(*request.Body.Description))
+	}
+	err = s.DB.UpdateAnalysis(s.DB.NoTxn(ctx), id, mutations...)
+	if err != nil {
+		return nil, oapierr.Internal("failed to update analysis", zap.Error(err))
+	}
```

Comment:
optional: To avoid loading the analysis twice, you could have `checkAnalysisAuthorization` optionally provide a transaction, and then use a `PutAnalysis`-type DB update instead.

Reply:
I'll skip this at the moment. If we have performance implications that are noticeable I'll obviously reconsider. My thinking is twofold: (a) keeping the authz code as simple as possible has complexity benefits (b) we only allow some fields to be changeable. I like that we expose that set of fields via `db.UpdateFns`, since it allows us to have an explicit API that shows the bizlogic which fields are writable. Put semantics blur this line, and can easily cause a situation where the bizlogic tries to set an field that isn't settable, and ends up surprised by it.